src/freertos_drivers/spiffs/SPIFFS.cxx

```diff
@@ -0,0 +1,537 @@
+        ffs_stat.type = SPIFFS_TYPE_DIR;
+        ffs_stat.obj_id = 0;
+        ffs_stat.size = 0;
+    }
+    else
+    {
+        ssize_t result = SPIFFS_stat(&fs_, path, &ffs_stat);
+
+        if (result < 0)
+        {
+            return -errno_translate(result);
+        }
+    }
+
+    memset(stat, 0, sizeof(*stat));
+    stat->st_ino = ffs_stat.obj_id;
+    stat->st_size = ffs_stat.size;
+    switch (ffs_stat.type)
+    {
+        default:
+            break;
+        case SPIFFS_TYPE_FILE:
+            stat->st_mode = S_IFREG;
+            break;
+        case SPIFFS_TYPE_DIR:
+            stat->st_mode = S_IFDIR | S_IXUSR | S_IXGRP | S_IXOTH;;
+            break;
+        case SPIFFS_TYPE_HARD_LINK:
+            stat->st_mode = S_IFLNK | S_IXUSR | S_IXGRP | S_IXOTH;;
+            break;
+        case SPIFFS_TYPE_SOFT_LINK:
+            stat->st_mode = S_IFLNK | S_IXUSR | S_IXGRP | S_IXOTH;;
+            break;
+    }
+
+    stat->st_mode |= S_IRUSR | S_IRGRP | S_IROTH |
+                     S_IWUSR | S_IWGRP | S_IWOTH;
+
+    return 0;
+}
+
+//
+// SPIFFS::fsync()
+//
+int SPIFFS::fsync(File *file)
+{
+    spiffs_file fd = file->privInt;
+    int result = SPIFFS_fflush(&fs_, fd);
+    if (result < 0)
+    {
+        return -errno_translate(result);
+    }
+
+    return 0;
+}
+
+//
+// SPIFFS::closedir()
+//
+int SPIFFS::closedir(File *file)
+{
+    OpenDir *dir = static_cast<OpenDir*>(file->priv);
+    int result = SPIFFS_closedir(&dir->dir_);
+
+    if (result == 0)
+    {
+        free(dir);
+        return 0;
+    }
+    else
+    {
+        return -errno_translate(result);
+    }
+}
+
+//
+// SPIFFS::opendir()
+//
+File *SPIFFS::opendir(File *file, const char *name)
+{
+    // special malloc is to reserve space for the full path name
+    OpenDir *dir = static_cast<OpenDir*>(malloc(sizeof(OpenDir) +
+                                                SPIFFS_OBJ_NAME_LEN +
+                                                strlen(this->name) + 1));
+
+    extern_lock(&fs_);
+    spiffs_DIR *result = SPIFFS_opendir(&fs_, name, &dir->dir_);
+    if (!result)
+    {
+        free(dir);
+        errno = errno_translate(fs_.err_code);
+    }
+    else
+    {
+        // no error occured
+        file->priv = dir;
+        file->dir = true;
+    }
+    extern_unlock(&fs_);
+
+    return file;
+}
+
+//
+// SPIFFS::readdir()
+//
+struct dirent *SPIFFS::readdir(File *file)
+{
+    OpenDir *dir = static_cast<OpenDir*>(file->priv);
+    spiffs_dirent dirent;
+
+    spiffs_dirent *result = SPIFFS_readdir(&dir->dir_, &dirent);
+
+    if (!result)
+    {
+        return nullptr;
+    }
+    else
+    {
+        dir->dirent_.d_ino = dirent.obj_id;
+        strcpy(dir->dirent_.d_name, this->name);
+        strcat(dir->dirent_.d_name, "/");
+        strcat(dir->dirent_.d_name, (char*)dirent.name);
+        return &dir->dirent_;
+    }
+}
+
+//
+// SPIFFS::errno_translate()
+//
+int SPIFFS::errno_translate(int spiffs_error)
+{
+    switch (spiffs_error)
```

Comment:
consider saving this value into a global variable that we can look at in the debugger if needed.

Reply:
Done.